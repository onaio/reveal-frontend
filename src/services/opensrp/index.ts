import { IncomingHttpHeaders } from 'http';
import { OPENSRP_API_BASE_URL } from '../../configs/env';
import store from '../../store';
import { getAccessToken } from '../../store/selectors';

/** allowed http methods */
type HTTPMethod = 'GET' | 'POST' | 'PUT';

/** get default HTTP headers for OpenSRP service */
export function getDefaultHeaders(
  accept: string = 'application/json',
  authorizationType: string = 'Bearer',
  contentType: string = 'application/json;charset=UTF-8'
): IncomingHttpHeaders {
  return {
    accept,
    authorization: `${authorizationType} ${getAccessToken(store.getState())}`,
    'content-type': contentType,
  };
}

/** interface to describe URL params object */
export interface URLParams {
  [key: string]: string | number | boolean;
}

/** params option type */
type paramsType = URLParams | null;

/** converts URL params object to string
 * @param {URLParams} obj - the object representing URL params
 * @returns {string} URL params as a string
 */
export function getURLParams(obj: URLParams | {}): string {
  return Object.entries(obj)
    .map(([key, val]) => `${key}=${val}`)
    .join('&');
}

/** converts filter params object to string
 * @param {URLParams} obj - the object representing filter params
 * @returns {string} filter params as a string
 */
export function getFilterParams(obj: URLParams | {}): string {
  return Object.entries(obj)
    .map(([key, val]) => `${key}:${val}`)
    .join(',');
}

/** get payload for fetch
 * @param {HTTPMethod} method - the HTTP method
 * @returns the payload
 */
export function getPayload(method: HTTPMethod) {
  return {
    headers: getDefaultHeaders() as HeadersInit,
    method,
  };
}

/** Get URL
 * @param {string} url - the url
 * @param {paramType} params - the url params object
 * @returns {string} the final url
 */
export function getURL(url: string, params: paramsType = null): string {
  let result = url;
  if (params) {
    result = `${result}?${getURLParams(params)}`;
  }
  return result;
}

/** The OpenSRP service class
 *
 * Sample usage:
 * -------------
 * const service = new OpenSRPService('the-endpoint');
 *
 * **To list all objects**: service.list()
 *
 * **To get one object**: service.read('the-object-identifier')
 *
 * **To create a new object**: service.create(theObject)
 *
 * **To update an object**: service.update(theObject)
 */
export class OpenSRPService {
  public baseURL: string;
  public endpoint: string;
  public generalURL: string;

  constructor(endpoint: string, baseURL: string = OPENSRP_API_BASE_URL) {
    this.endpoint = endpoint;
    this.baseURL = baseURL;
    this.generalURL = `${this.baseURL}${this.endpoint}`;
  }

  /** create method
   * Send a POST request to the general endpoint containing the new object data
   * Successful requests will result in a HTTP status 201 response with no body
   * @param {T} data - the data to be POSTed
   * @param {params} params - the url params object
   * @param {HTTPMethod} method - the HTTP method
   * @returns the object returned by API
   */
  public async create<T>(data: T, params: paramsType = null, method: HTTPMethod = 'POST') {
    const url = getURL(this.generalURL, params);
    const payload = {
      ...getPayload(method),
      'Cache-Control': 'no-cache',
      Pragma: 'no-cache',
      body: JSON.stringify(data),
    };
    const response = await fetch(url, payload);

    if (!response.ok || response.status !== 201) {
      throw new Error(
        `OpenSRPService create on ${this.endpoint} failed, HTTP status ${response.status}`
      );
    }

    return {};
  }

  /** read method
   * Send a GET request to the url for the specific object
   * @param {string|number} id - the identifier of the object
   * @param {params} params - the url params object
   * @param {HTTPMethod} method - the HTTP method
   * @returns the object returned by API
   */
  public async read(id: string | number, params: paramsType = null, method: HTTPMethod = 'GET') {
    const url = getURL(`${this.generalURL}/${id}`, params);
    const response = await fetch(url, getPayload(method));

    if (!response.ok) {
      throw new Error(
        `OpenSRPService read on ${this.endpoint} failed, HTTP status ${response.status}`
      );
    }

    return await response.json();
  }

  /** update method
   * Simply send the updated object as PUT request to the general endpoint URL
   * Successful requests will result in a HTTP status 200/201 response with no body
   * @param {T} data - the data to be POSTed
   * @param {params} params - the url params object
   * @param {HTTPMethod} method - the HTTP method
   * @returns the object returned by API
   */
  public async update<T>(data: T, params: paramsType = null, method: HTTPMethod = 'PUT') {
    const url = getURL(this.generalURL, params);
    const payload = {
      ...getPayload(method),
      'Cache-Control': 'no-cache',
      Pragma: 'no-cache',
      body: JSON.stringify(data),
    };
    const response = await fetch(url, payload);

    if (!response.ok) {
      throw new Error(
        `OpenSRPService update on ${this.endpoint} failed, HTTP status ${response.status}`
      );
    }

    return {};
  }

  /** list method
   * Send a GET request to the general API endpoint
   * @param {params} params - the url params object
   * @param {HTTPMethod} method - the HTTP method
   * @returns list of objects returned by API
   */
  public async list(params: paramsType = null, method: HTTPMethod = 'GET') {
    const url = getURL(this.generalURL, params);
    const response = await fetch(url, getPayload(method));

    if (!response.ok) {
      throw new Error(
        `OpenSRPService list on ${this.endpoint} failed, HTTP status ${response.status}`
      );
    }

    return await response.json();
  }
}
