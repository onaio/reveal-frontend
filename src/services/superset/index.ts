import reducerRegistry from '@onaio/redux-reducer-registry';
import superset, { SupersetConnectorConfig } from '@onaio/superset-connector';
import { OPENSRP_OAUTH_STATE, SUPERSET_API_BASE, SUPERSET_API_ENDPOINT } from '../../configs/env';
import { FlexObject } from '../../helpers/utils';
import store from '../../store';
import supersetReducer, {
  authorizeSuperset,
  isAuthorized,
  reducerName as supersetReducerName,
} from '../../store/ducks/superset';
import { getAccessToken, getOauthProviderState } from '../../store/selectors';

// /** register the superset reducer */
reducerRegistry.register(supersetReducerName, supersetReducer);

/** middleware for fetching from Superset */
export const fetchMiddleware = (res: { [key: string]: any }) => {
  return res;
};

/** callback for fetching from Superset */
export const fetchCallback = (parsedResponse: Array<{ [key: string]: any }>) => {
  const sliceData = superset.processData(parsedResponse);
  return sliceData;
};

/** this function completes the authZ process */
export const completeAuthZ = async (result: { [key: string]: any }) => {
  if (result.status === 200) {
    store.dispatch(authorizeSuperset(true));
  } else {
    store.dispatch(authorizeSuperset(false));
  }
};

/** generic async function for fetching from superset */
const supersetFetch = async (
  sliceId: string,
  params: FlexObject | null = null,
  callback: typeof fetchCallback = fetchCallback,
  middleware: typeof fetchMiddleware = fetchMiddleware
) => {
  const config: SupersetConnectorConfig = {
    base: SUPERSET_API_BASE,
    endpoint: SUPERSET_API_ENDPOINT,
    extraPath: sliceId,
    /** the provider should be the name of the provider
     * to enable the use of multiple providers, we store the provider name in the
     * `state` parameter so that we can use it here
     */
    provider: getOauthProviderState(store.getState()) || OPENSRP_OAUTH_STATE,
    token: '',
  };

  const accessToken = getAccessToken(store.getState());
  if (accessToken) {
    config.token = accessToken;
  }

  if (params) {
    config.params = `form_data=${JSON.stringify(params)}`;
  }

  const isSupersetAuthorized = isAuthorized(store.getState());

  if (isSupersetAuthorized === true) {
    return superset.api.doFetch(config, middleware).then(callback);
  } else {
    return superset.authZ(config, (result: { [key: string]: any }) => {
      return completeAuthZ(result)
        .then(() => {
          const isSupersetAuthorizedYet = isAuthorized(store.getState());
          if (isSupersetAuthorizedYet === true) {
            return superset.api.doFetch(config, middleware).then(callback);
          }
        })
        .catch(() => {
          return false;
        });
    });
  }
};

export default supersetFetch;
