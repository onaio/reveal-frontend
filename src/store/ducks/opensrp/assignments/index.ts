/** Assignments redux module */
import { Dictionary } from '@onaio/utils';
import { get } from 'lodash';
import { Store } from 'redux';
import { AnyAction } from 'redux';
import SeamlessImmutable from 'seamless-immutable';
import { setDefaultValues } from '../../../../helpers/utils';

/** The reducer name */
export const reducerName = 'assignments';

/** interface for a Assignment object */
export interface Assignment {
  jurisdiction: string;
  organization: string;
  plan: string;
  fromDate: string;
  toDate: string;
}

// action interfaces

/** action type for action that adds Assignments to store */
export const ASSIGNMENTS_FETCHED = 'src/store/ducks/assignments/reducer/ASSIGNMENTS_FETCHED';
/** action type for REMOVE_TEAMS action */
export const REMOVE_ASSIGNMENTS = 'src/store/ducks/assignments/reducer/REMOVE_ASSIGNMENTS';

/** interface for Assignments fetched action */
interface FetchAssignmentsAction extends AnyAction {
  assignmentsByPlanId: { [key: string]: Assignment[] };
  type: typeof ASSIGNMENTS_FETCHED;
}

/** interface for action that removes assignments from store */
interface RemoveAssignmentsAction extends AnyAction {
  assignmentsByPlanId: { [key: string]: Assignment[] };
  type: typeof REMOVE_ASSIGNMENTS;
}

/** single type for all action types */
type AssignmentActionTypes = FetchAssignmentsAction | RemoveAssignmentsAction | AnyAction;

/** interface describing the assignmentsByPlanId Store */
interface AssignmentsByPlanId {
  [key: string]: Assignment[];
}
/** interface for Assignments state in store */
interface AssignmentsStoreState {
  assignmentsByPlanId: AssignmentsByPlanId | {};
}

// immutable assignments state in dux
export type ImmutableAssignmentsStoreState = AssignmentsStoreState &
  SeamlessImmutable.ImmutableObject<AssignmentsStoreState>;

/** initial state for Assignments records in store */
const initialOrgsStoreState: ImmutableAssignmentsStoreState = SeamlessImmutable({
  assignmentsByPlanId: {},
});

/** the Assignment reducer function */
export default function reducer(state = initialOrgsStoreState, action: AssignmentActionTypes) {
  switch (action.type) {
    case ASSIGNMENTS_FETCHED:
      return SeamlessImmutable({
        ...state,
        assignmentsByPlanId: {
          ...state.assignmentsByPlanId,
          ...action.assignmentsByPlanId,
        },
      });
    case REMOVE_ASSIGNMENTS:
      return SeamlessImmutable({
        ...state,
        assignmentsByPlanId: action.assignmentsByPlanId,
      });
    default:
      return state;
  }
}

/** action to remove assignments form store */
export const removeAssignmentsAction: RemoveAssignmentsAction = {
  assignmentsByPlanId: {},
  type: REMOVE_ASSIGNMENTS,
};

// action creators

/** creates action to add fetched assignments to store
 * @param {Assignment []} assignmentsList - array of assignments to be added to store
 *
 * @returns {FetchAssignmentsAction} - action with assignments payload that is added to store
 */
export const fetchAssignments = (assignmentsList: Assignment[]): FetchAssignmentsAction => {
  // const assignmentsByPlanId: AssignmentsByPlanId = {};
  const defaultArrayAssignmentHandler: ProxyHandler<object> = {
    /**
     * The methods inside the handler object are called traps for our case we use a get trap
     * @param {AssignmentsByPlanId} target
     * @param {string} plan planId the assignmentsByPlanId will be updated by empty arrays keyed by the planId
     */
    get: (target: AssignmentsByPlanId, plan: string) => {
      /**
       * Check if the planId already exists then return the target else add the key with an empty array value
       */
      return setDefaultValues(target, plan);
    },
  };

  let assignmentsByPlanId: AssignmentsByPlanId = {};
  /**
   * Todo:Research on a solution  for Typescript can't infer types when using Proxy
   */
  /**
   * A JavaScript Proxy is an object that wraps another object (target) and intercepts the fundamental operations of the target object
   * @param {AssignmentsByPlanId} assignmentsByPlanId target(is an object to wrap) initial value is an empty object
   * @param {ProxyHandler<object>} defaultArrayAssignmentHandler  object that contains methods to control the behaviors of the target
   */
  const assignmentsByPlanIdProxy = new Proxy(
    assignmentsByPlanId,
    defaultArrayAssignmentHandler
  ) as Dictionary;
  for (const assignment of assignmentsList) {
    assignmentsByPlanId = assignmentsByPlanIdProxy[assignment.plan];
    assignmentsByPlanId[assignment.plan].push(assignment);
  }
  return {
    assignmentsByPlanId,
    type: ASSIGNMENTS_FETCHED,
  };
};

/** creates action to reset plan
 * @param {AssignmentsByPlanId} assignmentsByPlanId object with updated assignment arrays, keyed by planId
 * @returns {FetchAssignmentsAction}
 */
export const resetPlanAssignments = (
  assignmentsByPlanId: AssignmentsByPlanId = {}
): FetchAssignmentsAction => {
  return {
    assignmentsByPlanId,
    type: ASSIGNMENTS_FETCHED,
  };
};

// selectors

/** get assignments as an object where their ids are the keys and the objects
 * the values
 * @param {Partial<Store>} state - Portion of the store
 *
 * @return {[key: string]: Assignment}
 */
export function getAssignmentsByPlanId(state: Partial<Store>): { [key: string]: Assignment[] } {
  return (state as any)[reducerName].assignmentsByPlanId;
}

/** Get all assignments as an array
 * @param {Partial<Store>} state - Part of the redux store
 * @param {string} planId - The id of the plan of which to return associated Assignments
 *
 * @return {Assignment []} - all assignments by planId in store as an array
 */
export function getAssignmentsArrayByPlanId(state: Partial<Store>, planId: string): Assignment[] {
  return get(getAssignmentsByPlanId(state), planId) || [];
}

/** Get all assignments by plan id and by jurisdiction id
 * @param {Partial<Store>} state - Part of the redux store
 * @param {string} planId - The id of the plan of which to return associated Assignments
 * @param {string} jurisdictionId = The id of the jurisdiction of which to return associated Assignments
 *
 * @return {Assignment []} - all assignments by planId in store as an array
 */
export function getAssignmentsArrayByPlanIdByJurisdictionId(
  state: Partial<Store>,
  planId: string,
  jurisdictionId: string
): Assignment[] {
  return getAssignmentsArrayByPlanId(state, planId).filter(
    (a: Assignment) => a.jurisdiction === jurisdictionId
  );
}
