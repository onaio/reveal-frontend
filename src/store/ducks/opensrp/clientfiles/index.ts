import { keyBy, values } from 'lodash';
import { AnyAction, Store } from 'redux';
import { createSelector } from 'reselect';
import SeamlessImmutable from 'seamless-immutable';

/** The reducer name */
export const reducerName = 'clientfiles';

/** Interface for file json object */
export interface File {
  identifier: string; // the unique identifier of the file
  fileLength?: number; // the length of the file (lines / rows)
  fileName: string; // the name of the file
  fileSize?: string; // the of the file
  uploadDate: string; // the date of the latest file update
  providerID: string; // the username of the file creator
  url: string; // download location of the file
}

// actions

/** action type for fetching clients */
export const FILES_FETCHED = 'opensrp/reducer/clients/FILES_FETCHED';
/** action type for removing clients */
export const REMOVE_FILES = 'opensrp/reducer/clients/REMOVE_FILES';

/** interface action to add Files to store */
export interface FetchFilesAction extends AnyAction {
  overwrite: boolean;
  filesById: { [key: string]: File };
  type: typeof FILES_FETCHED;
}

/** Interface for removeFilesAction */
export interface RemoveFilesAction extends AnyAction {
  filesById: {};
  type: typeof REMOVE_FILES;
}

/** Create type for file reducer actions */
export type FileActionTypes = FetchFilesAction | RemoveFilesAction | AnyAction;

// action Creators

/** Fetch files action creator
 * @param {File[]} clientsList - clients array to add to store
 * @param {boolean} overwrite - whether to replace the records in store for clients
 * @return {FetchFileAction} - an action to add clients to redux store
 */
export const fetchFiles = (
  filesList: File[] = [],
  overwrite: boolean = false
): FetchFilesAction => ({
  filesById: keyBy(filesList, (file: File) => file.identifier),
  overwrite,
  type: FILES_FETCHED,
});

// actions

/** removeFilesAction action */
export const removeFilesAction = {
  filesById: {},
  type: REMOVE_FILES,
};

// The reducer

/** interface for files state in redux store */
export interface FileState {
  filesById: { [key: string]: File } | {};
}

/** Create an immutable Files state */
export type ImmutableFileState = FileState & SeamlessImmutable.ImmutableObject<FileState>;

/** initial practitioners-state state */
export const initialState: ImmutableFileState = SeamlessImmutable({
  filesById: {},
});

/** the Files reducer function */
export default function reducer(
  state: ImmutableFileState = initialState,
  action: FileActionTypes
): ImmutableFileState {
  switch (action.type) {
    case FILES_FETCHED:
      const filesToPut = action.overwrite
        ? { ...action.filesById }
        : { ...state.filesById, ...action.filesById };
      return SeamlessImmutable({
        ...state,
        filesById: filesToPut,
      });
    case REMOVE_FILES: {
      return SeamlessImmutable({
        ...state,
        filesById: action.filesById,
      });
    }
    default:
      return state;
  }
}

// Selectors

/** Get all Files keyed by File.identifier
 * @param {Partial<Store>} state - Portion of the store
 * @returns {[key: string]: File} filesById
 */
export const getFilesById = (state: Partial<Store>): { [key: string]: File } => {
  return (state as any)[reducerName].filesById;
};

/** Get all Files as an array of File objects
 * @param {Partial<Store>} state - the redux store
 * @returns {File[]} - an array of File objects
 */
export const getFilesArray = (state: Partial<Store>): File[] => {
  return values(getFilesById(state) || {});
};

/** This interface represents the structure of files filter options/params */
export interface FileFilters {
  fileName?: string /** file name */;
}

/**
 * Gets file name from Filters
 * @param state - the redux store
 * @param props - the file filters object
 */
export const getFileName = (_: Partial<Store>, props: FileFilters) => props.fileName;

/**
 * Gets an array of files filtered by name
 * @param {Partial<Store>} state - the redux store
 * @param {FileFilters} props - the  files filters object
 */
export const getFilesArrayByName = () =>
  createSelector([getFilesArray, getFileName], (files, name) =>
    name ? files.filter(file => file.fileName?.toLowerCase().includes(name.toLowerCase())) : files
  );

/**
 * Returns a selector that gets an array of file objects filtered by one or all
 * of the following:
 *    - file_name
 *
 * These filter params are all optional and are supplied via the prop parameter.
 *
 * This selector is meant to be a memoized replacement for getFilesById.
 *
 * To use this selector, do something like:
 *    const filesArraySelector = makeFilesArraySelector();
 *
 * @param {Partial<Store>} state - the redux store
 * @param {FileFilters} props - the files filters object
 */
export const makeFilesArraySelector = () => {
  return createSelector([getFilesArrayByName()], file1 => file1);
};
