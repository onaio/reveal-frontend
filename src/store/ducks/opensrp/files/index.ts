import { keyBy, values } from 'lodash';
import { AnyAction, Store } from 'redux';
import SeamlessImmutable from 'seamless-immutable';

/** The reducer name */
export const reducerName = 'files';

/** Interface for file json object */
export interface File {
  identifier: string; // the unique identifier of the file
  fileLength: number; // the length of the file (lines / rows)
  fileName: string; // the name of the file
  fileSize: string; // the of the file
  lastUpdated: string; // the date of the latest file update
  owner: string; // the username of the file creator
  url: string; // download location of the file
}

// actions

/** action type for fetching students */
export const FILES_FETCHED = 'opensrp/reducer/students/FILES_FETCHED';
/** action type for removing students */
export const REMOVE_FILES = 'opensrp/reducer/students/REMOVE_FILES';

/** interface action to add Files to store */
export interface FetchFilesAction extends AnyAction {
  overwrite: boolean;
  filesById: { [key: string]: File };
  type: typeof FILES_FETCHED;
}

/** Interface for removeFilesAction */
export interface RemoveFilesAction extends AnyAction {
  filesById: {};
  type: typeof REMOVE_FILES;
}

/** Create type for file reducer actions */
export type FileActionTypes = FetchFilesAction | RemoveFilesAction | AnyAction;

// action Creators

/** Fetch files action creator
 * @param {File[]} studentsList - students array to add to store
 * @param {boolean} overwrite - whether to replace the records in store for students
 * @return {FetchFileAction} - an action to add students to redux store
 */
export const fetchFiles = (
  filesList: File[] = [],
  overwrite: boolean = false
): FetchFilesAction => ({
  filesById: keyBy(filesList, (file: File) => file.identifier),
  overwrite,
  type: FILES_FETCHED,
});

// actions

/** removeFilesAction action */
export const removeFilesAction = {
  filesById: {},
  type: REMOVE_FILES,
};

// The reducer

/** interface for files state in redux store */
export interface FileState {
  filesById: { [key: string]: File } | {};
}

/** Create an immutable Files state */
export type ImmutableFileState = FileState & SeamlessImmutable.ImmutableObject<FileState>;

/** initial practitioners-state state */
export const initialState: ImmutableFileState = SeamlessImmutable({
  filesById: {},
});

/** the Files reducer function */
export default function reducer(
  state: ImmutableFileState = initialState,
  action: FileActionTypes
): ImmutableFileState {
  switch (action.type) {
    case FILES_FETCHED:
      const filesToPut = action.overwrite
        ? { ...action.filesById }
        : { ...state.filesById, ...action.filesById };
      return SeamlessImmutable({
        ...state,
        filesById: filesToPut,
      });
    case REMOVE_FILES: {
      return SeamlessImmutable({
        ...state,
        filesById: action.filesById,
      });
    }
    default:
      return state;
  }
}

// Selectors

/** Get all Files keyed by File.identifier
 * @param {Partial<Store>} state - Portion of the store
 * @returns {[key: string]: File} filesById
 */
export const getFilesById = (state: Partial<Store>): { [key: string]: File } => {
  return (state as any)[reducerName].filesById;
};

/** Get all Files as an array of File objects
 * @param {Partial<Store>} state - the redux store
 * @returns {File[]} - an array of File objects
 */
export const getFilesArray = (state: Partial<Store>): File[] => {
  return values(getFilesById(state));
};
