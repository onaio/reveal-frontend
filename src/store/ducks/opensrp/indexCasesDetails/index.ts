import {
  fetchActionCreatorFactory,
  getItemsByIdFactory,
  reducerFactory,
  removeActionCreatorFactory,
} from '@opensrp/reducer-factory';
import intersect from 'fast_array_intersect';
import { values } from 'lodash';
import { Store } from 'redux';
import { createSelector } from 'reselect';

/** reducer name for the Item module */
export const reducerName = 'indexCasesDetails';

export interface IndexCasesDetails {
  action_code: string;
  case_classification: string;
  case_number: string;
  date_of_diagnosis: string;
  event_id: string;
  goal_id: string;
  jurisdiction_id: string;
  id: string;
  plan_id: string;
  task_id: string;
}

/** Item Reducer */
const reducer = reducerFactory<IndexCasesDetails>(reducerName);
export default reducer;

// action
/** actionCreator returns action to to add Item records to store */
export const fetchIndexCaseDetails = fetchActionCreatorFactory<IndexCasesDetails>(
  reducerName,
  'id'
);
export const removeIndexCaseDetailsAction = removeActionCreatorFactory(reducerName);

// selectors
export const getAllIndexCaseDetailsByIds = getItemsByIdFactory<IndexCasesDetails>(reducerName);

/** This interface represents the structure of index case details filter options/params */
export interface IndexCaseFilters {
  goal_id?: string /** goal id */;
  jurisdiction_id?: string /** jurisdiction id */;
  plan_id?: string /* plan id */;
  task_id?: string /** task id */;
}

/** get plan id
 * Gets plan id from IndexCaseFilters
 * @param state - the redux store
 * @param props - the index cases filters object
 */
export const getPlanId = (_: Partial<Store>, props: IndexCaseFilters) => props.plan_id;

/** get jurisdiction id
 * Gets jurisdiction id from IndexCaseFilters
 * @param state - the redux store
 * @param props - the index cases filters object
 */
export const getJurisdictionId = (_: Partial<Store>, props: IndexCaseFilters) =>
  props.jurisdiction_id;

/** get goal id
 * Gets goal id from IndexCaseFilters
 * @param state - the redux store
 * @param props - the index cases filters object
 */
export const getGoalId = (_: Partial<Store>, props: IndexCaseFilters) => props.goal_id;

/** get task id
 * Gets task id from IndexCaseFilters
 * @param state - the redux store
 * @param props - the index cases filters object
 */
export const getTaskId = (_: Partial<Store>, props: IndexCaseFilters) => props.task_id;

/** indexCasesArrayBaseSelector select an array of all index cases
 * @param state - the redux store
 */
export const indexCasesArrayBaseSelector = (state: Partial<Store>): IndexCasesDetails[] =>
  values(getAllIndexCaseDetailsByIds(state) || {});

/** getIndexCaseArrayByPlanId
 * Gets an array of index cases objects filtered by plan id
 * @param  state - the redux store
 * @param  props - the index cases filters object
 */
export const getIndexCasesArrayByPlanId = () =>
  createSelector([indexCasesArrayBaseSelector, getPlanId], (indexCases, planId) => {
    return planId ? indexCases.filter(indexCase => indexCase.plan_id === planId) : indexCases;
  });

/** getIndexCaseArrayByJurisdictionId
 * Gets an array of index cases objects filtered by jurisdiction id
 * @param  state - the redux store
 * @param  props - the index cases filters object
 */
export const getIndexCasesArrayByJurisdictionId = () =>
  createSelector([indexCasesArrayBaseSelector, getJurisdictionId], (indexCases, jurisdiction) => {
    return jurisdiction
      ? indexCases.filter(indexCase => indexCase.jurisdiction_id === jurisdiction)
      : indexCases;
  });

/** getIndexCasesArrayByGoalId
 * Gets an array of index cases objects filtered by goal id
 * @param  state - the redux store
 * @param  props - the index cases filters object
 */
export const getIndexCasesArrayByGoalId = () =>
  createSelector([indexCasesArrayBaseSelector, getGoalId], (indexCases, goal) => {
    return goal ? indexCases.filter(indexCase => indexCase.goal_id === goal) : indexCases;
  });

/** getIndexCasesArrayByTaskId
 * Gets an array of index cases objects filtered by task id
 * @param  state - the redux store
 * @param  props - the index cases filters object
 */
export const getIndexCasesArrayByTaskId = () =>
  createSelector([indexCasesArrayBaseSelector, getTaskId], (indexCases, task) => {
    return task ? indexCases.filter(indexCase => indexCase.task_id === task) : indexCases;
  });

/** makeIndexCasesArraySelector
 * Returns a selector that gets an array of index cases objects filtered by one or all
 * of the following:
 *    - plan id
 *    - jurisdiction id
 *    - goal id
 *    - task id
 *
 * These filter params are all optional and are supplied via the prop parameter.
 *
 * To use this selector, do something like:
 *    const indexCasesArraySelector = makeIndexCasesArraySelector();
 *
 * @param  state - the redux store
 * @param {IndexCaseFilters} props - the index cases filters object
 */
export const makeIndexCasesArraySelector = () => {
  return createSelector(
    [
      getIndexCasesArrayByPlanId(),
      getIndexCasesArrayByJurisdictionId(),
      getIndexCasesArrayByGoalId(),
      getIndexCasesArrayByTaskId(),
    ],
    (case1, case2, case3, case4) => intersect([case1, case2, case3, case4], JSON.stringify)
  );
};
