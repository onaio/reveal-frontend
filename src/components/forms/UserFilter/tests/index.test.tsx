import { mount, shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import flushPromises from 'flush-promises';
import { createBrowserHistory } from 'history';
import React from 'react';
import { Router } from 'react-router';
import { UserSelectFilter } from '..';
import { OpenSRPService } from '../../../../services/opensrp';
import { users } from '../../PractitionerForm/UserIdSelect/tests/fixtures';

// tslint:disable-next-line:no-var-requires
const fetch = require('jest-fetch-mock');

const history = createBrowserHistory();

describe('src/components/forms/FilterForm', () => {
  it('renders without crashing', async () => {
    fetch.once(JSON.stringify(users.length)).once(JSON.stringify(users));
    const props = {
      serviceClass: OpenSRPService,
    };
    const wrapper = shallow(
      <Router history={history}>
        <UserSelectFilter {...props} />
      </Router>
    );
    // tslint:disable-next-line:promise-must-complete
    await new Promise<any>(resolve => new Promise<any>(resolve));
    wrapper.update();
    wrapper.unmount();
  });

  it('renders correctly', async () => {
    fetch.once(JSON.stringify(users.length)).once(JSON.stringify(users));
    const props = {
      serviceClass: OpenSRPService,
    };
    const wrapper = mount(
      <Router history={history}>
        <UserSelectFilter {...props} />
      </Router>
    );
    // tslint:disable-next-line:promise-must-complete
    await new Promise<any>(resolve => new Promise<any>(resolve));
    wrapper.update();
    const inputSelect = wrapper.find('input');
    expect(toJson(inputSelect)).toMatchSnapshot('Selector Input');
  });

  it('calls to fetch', async () => {
    fetch.once(JSON.stringify(users.length)).once(JSON.stringify(users));

    const props = {
      serviceClass: OpenSRPService,
    };
    mount(
      <Router history={history}>
        <UserSelectFilter {...props} />
      </Router>
    );
    // tslint:disable-next-line:promise-must-complete
    await new Promise<any>(resolve => new Promise<any>(resolve));
    await flushPromises();
    const calls = [
      [
        'https://reveal-stage.smartregister.org/opensrp/rest/user/count',
        {
          headers: {
            accept: 'application/json',
            authorization: 'Bearer null',
            'content-type': 'application/json;charset=UTF-8',
          },
          method: 'GET',
        },
      ],
      [
        'https://reveal-stage.smartregister.org/opensrp/rest/user?page_size=1000&source=Keycloak&start_index=0',
        {
          headers: {
            accept: 'application/json',
            authorization: 'Bearer null',
            'content-type': 'application/json;charset=UTF-8',
          },
          method: 'GET',
        },
      ],
      [
        'https://reveal-stage.smartregister.org/opensrp/rest/user/count',
        {
          headers: {
            accept: 'application/json',
            authorization: 'Bearer null',
            'content-type': 'application/json;charset=UTF-8',
          },
          method: 'GET',
        },
      ],
      [
        'https://reveal-stage.smartregister.org/opensrp/rest/user?page_size=1000&source=Keycloak&start_index=0',
        {
          headers: {
            accept: 'application/json',
            authorization: 'Bearer null',
            'content-type': 'application/json;charset=UTF-8',
          },
          method: 'GET',
        },
      ],
    ];
    expect(fetch.mock.calls).toEqual(calls);
  });
  it('calls onchangeHandler callback correctly with correct arguments', async () => {
    fetch.once(JSON.stringify(users.length)).once(JSON.stringify(users));
    const mock: any = jest.fn();
    const props = {
      onChangeHandler: mock,
      serviceClass: OpenSRPService,
    };
    const wrapper = mount(
      <Router history={history}>
        <UserSelectFilter {...props} />
      </Router>
    );

    await flushPromises();
    wrapper.update();

    (wrapper.find('Select').instance() as any).selectOption({
      label: 'Drake.Ramole',
      value: '0259c0bc-78a2-4284-a7a9-d61d0005djae',
    });
    wrapper.update();

    // what is the onchangeHandler called with
    expect(mock.mock.calls).toEqual([
      [
        {
          label: 'Drake.Ramole',
          value: '0259c0bc-78a2-4284-a7a9-d61d0005djae',
        },
      ],
    ]);
  });
});
