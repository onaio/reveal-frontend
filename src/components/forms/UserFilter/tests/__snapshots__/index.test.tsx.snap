// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src/components/forms/FilterForm renders correctly: Selector Input 1`] = `
<input
  aria-autocomplete="list"
  autoCapitalize="none"
  autoComplete="off"
  autoCorrect="off"
  disabled={false}
  id="react-select-2-input"
  onBlur={[Function]}
  onChange={[Function]}
  onFocus={[Function]}
  spellCheck="false"
  style={
    Object {
      "background": 0,
      "border": 0,
      "boxSizing": "content-box",
      "color": "inherit",
      "fontSize": "inherit",
      "label": "input",
      "opacity": 1,
      "outline": 0,
      "padding": 0,
      "width": "2px",
    }
  }
  tabIndex="0"
  type="text"
  value=""
/>
`;

exports[`src/components/forms/FilterForm renders correctly: label and input 1`] = `
Object {
  "children": <label>
    user
  </label>,
  "className": "col-sm-4",
}
`;

exports[`src/components/forms/FilterForm renders correctly: label and input 2`] = `
Object {
  "children": <UserIdSelect
    ReactSelectDefaultValue={
      Object {
        "label": "Select username",
        "value": "Select username",
      }
    }
    className=""
    onChangeHandler={[Function]}
    serviceClass={[Function]}
    showPractitioners={true}
    userNameAsValue={true}
  />,
  "className": "col-sm-8",
}
`;
